tests/samplers_tests/test_cmaes.py

```diff
@@ -400,28 +402,147 @@ def test_restore_optimizer_keeps_backward_compatibility() -> None:
     assert n_restarts == 1
 
 
-@pytest.mark.filterwarnings("ignore::optuna.exceptions.ExperimentalWarning")
-def test_restore_optimizer_from_substrings() -> None:
-    sampler = optuna.samplers.CmaEsSampler()
-    optimizer = CMA(np.zeros(10), sigma=1.3)
-    optimizer_str = pickle.dumps(optimizer).hex()
+@pytest.mark.parametrize("sampler_opts", [{}, {"use_separable_cma": True}, {"with_margin": True}])
+def test_restore_optimizer_from_substrings(sampler_opts: Dict[str, Any]) -> None:
+    popsize = 8
+    sampler = optuna.samplers.CmaEsSampler(popsize=popsize, **sampler_opts)
+    optimizer, n_restarts = sampler._restore_optimizer([])
+    assert optimizer is None
+    assert n_restarts == 0
 
-    system_attrs: Dict[str, Any] = sampler._split_optimizer_str(optimizer_str)
-    assert len(system_attrs) > 1
-    system_attrs["cma:n_restarts"] = 1
+    def objective(trial: optuna.Trial) -> float:
+        x1 = trial.suggest_float("x1", -10, 10, step=1)
+        x2 = trial.suggest_float("x2", -10, 10)
+        return x1**2 + x2**2
```

Comment:
This objective function appears so many times, how about defining this function on the top of this test?

Reply:
@gen740 Thank you for your review!

I understand your motivation. Obviously, this is not optimal in DRY mind. It's a matter of preference, but reusing the same objective function in multiple test cases often made tests complex. (e.g. `prepare_study_with_trials` function, which is used in visualization tests, contains a lot of implicit specifications and lots of unit tests depends on it. See https://github.com/optuna/optuna/pulls?q=is%3Apr+author%3Ac-bata+prepare_study_with_trials for details.)

So if it's a not your strong opinion, I suggest keeping this function not DRY. What do you think?

Reply:
I see. Indeed, reusing an objective function is sometimes hard to read...
I think that's fine to define for each test.